### PR TITLE
feat: installer

### LINKED ISSUE #471: Wanted: installer
```
deno run -A https://deno.land/std/installer/install.ts
```
a program which installs itself into your $PATH. It will walk your $PATH, find a writable directory, writing a new shell script into it called `$PATH/deno_install`

The shell script will look like this:
```sh
#!/bin/sh
deno run -A https://deno.land/std/installer/bin.ts $@
```
So now people can run this:
```
deno_install file_server https://deno.land/std/http/file_server.ts --allow-write --allow-net
```
This will create a new executable shell script called `$PATH/file_server`, which would have the following contents:
```sh
#!/bin/sh
deno run --allow-write --allow-net https://deno.land/std/http/file_server.ts $@
```

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #471 

This PR builds on #488 by @syumai

I've managed to remove call to `wget`

### CODE REVIEW

**Comment by bartlomieju** [comment]
```shellsession
20:07 $ deno -A ./installer/mod.ts https://deno.land/std/http/file_server.ts
[1/1] Compiling file:///Users/biwanczuk/dev/deno_std/installer/mod.ts
⚠️  file_server is already installed, do you want to overwrite it? [yN]
y

Downloading: https://deno.land/std/http/file_server.ts

ℹ️  Detected shebang:

   #!/usr/bin/env deno --allow-net

   Requested permissions:

	--allow-net

⚠️  Grant? [yN]
y

✅  Successfully installed file_server.

ℹ️  Add ~/.deno/bin to PATH
   echo 'export PATH="$HOME/.deno/bin:$PATH"' >> ~/.bashrc # change this to your shell
```

**Comment by ry** [comment]
```
ℹ️  Detected shebang:

   #!/usr/bin/env deno --allow-net

   Requested permissions:

	--allow-net

⚠️  Grant? [yN]
y
```
How does an installer request permissions?

**Comment by bartlomieju** [comment]
> How does an installer request permissions?

Not sure what you mean

**Comment by ry** [comment]
I mean if a user installs something like this
```
deno install --allow-net https://foo.bar/blah.ts
```
Then there is no prompt necessary because the user has already explicitly given permission.

So I don't understand when `Requested permissions: --allow-net` is displayed?

**Comment by ry** [comment]
Ooh it's based on the shebang? Hmm

Can we omit this feature for V1 ? shebang is quite unix specific, and it seems strange that it would have meaning on windows.

**Comment by bartlomieju** [comment]
> Ooh it's based on the shebang?

Yes, @syumai implemented it.

> Can we omit this feature for V1 ? shebang is quite unix specific, and it seems strange that it would have meaning on windows.

Sure, I wasn't sure about it anyway.

**Comment by bartlomieju** [comment]
@ry this should be ready for another review - I haven't tested it on Windows

**Comment by bartlomieju** [comment]
Okay, there's one more important feature to add:
```
deno_install file_server https://deno.land/std/http/file_server.ts --allow-write --allow-net
```
First argument (`file_server`) denotes how installed script should be named - currently it is done automatically by stripping name from installed script. I'll update that shortly

File: installer/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+Install remote or local script as executables.
+
+````
+## Installation
+
+`installer` can be install using iteself:
+
+```sh
+deno -A https://deno.land/std/install/deno_install.ts deno_install https://deno.land/std/install/deno_install.ts -A
```
Comment (ry) [inline installer/README.md]: s/install/installer

File: installer/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# deno_install
```
Comment (ry) [inline installer/README.md]: Let's use the name "installer" throughout to distinguish it from https://github.com/denoland/deno_install

Comment (bartlomieju) [reply installer/README.md]: Ah, good catch, you're right!

File: installer/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  Install remote or local script as executables.
+
+USAGE:
+  deno https://deno.land/std/installer/mod.ts EXE_NAME SCRIPT_URL [FLAGS...]
+
+ARGS:
+  EXE_NAME  Name for executable
+  SCRIPT_URL  Local or remote URL of script to install
+  [FLAGS...]  List of flags for script, both Deno permission and script specific flag can be used.
+```
```
Comment (ry) [inline installer/README.md]: It seems when I run it with no arguments, there's no error message or help text:
```
~/src/deno_std> deno -A installer/deno_installer.ts
[1/1] Compiling file:///Users/rld/src/deno_std/installer/deno_installer.ts
~/src/deno_std>
```

Comment (bartlomieju) [reply installer/README.md]: Yikes, can we just drop `installer/deno_installer.ts` and leave `installer/mod.ts`?

**Comment by ry** [comment]
Cool. It's working for me.

So this doesn't need to be solved now, but something to think about. I did this:
```
> deno -A ./installer/mod.ts file_server2 https://deno.land/std/http/file_server.ts -A
Downloading: https://deno.land/std/http/file_server.ts

✅ Successfully installed file_server2.
> which file_server2
/Users/rld/.deno/bin/file_server2
> file_server2 ../deno/website/
HTTP server listening on http://0.0.0.0:4500/
[2019-06-14 15:21:32] "GET / HTTP/1.1" 200
```
That worked correctly. 

But now I want to perhaps re-download the file_server2 program. I want to be able to do
```
file_server2 --reload
```
and have it operate as if I did
```
deno --reload https://deno.land/std/http/file_server.ts
```
but due to the placement of `$@` it won't pick up the `--reload` flag.

**Comment by bartlomieju** [comment]
@ry yep, I know about this caveat, but haven't figured it out yet.

**Comment by ry** [comment]
I mean it would be nice if for all deno programs, the argument ordering wasn't so necessary. It would be great if you could do "deno https://deno.land/std/http/file_server.ts  --reload"

**Comment by bartlomieju** [comment]
> I mean it would be nice if for all deno programs, the argument ordering wasn't so necessary. It would be great if you could do "deno https://deno.land/std/http/file_server.ts --reload"

I'll see what we can do about that.

> Is there any other changes you want to make before landing?
I added `uninstall` function but it's currently not exposed. I'm fine with landing as is.

**Comment by syumai** [comment]
LGTM to me too. Thanks @bartlomieju and @ry !

**Pull request merged**